### PR TITLE
Adds a class that can build a points processing pipeline out of a Lua configuration.

### PR DESCRIPTION
**Pull request by SirVer**

### CODE REVIEW

File: cartographer/io/points_processor_pipeline_builder.cc [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                                                  FactoryFunction factory) {
+  CHECK(factories_.count(name) == 0) << "A points processor with named '"
+                                     << name
+                                     << "' has already been registered.";
+  factories_[name] = factory;
+}
+
+std::vector<std::unique_ptr<PointsProcessor>>
+PointsProcessorPipelineBuilder::CreatePipeline(
+    common::LuaParameterDictionary* dictionary) const {
```
Comment (damonkohler) [inline cartographer/io/points_processor_pipeline_builder.cc]: const

File: cartographer/io/xyz_writing_points_processor.cc [hunk 2]
```diff
@@ -21,6 +22,13 @@ XyzWriterPointsProcessor::XyzWriterPointsProcessor(const string& filename,
   file_ << std::setprecision(6);
 }
 
+std::unique_ptr<XyzWriterPointsProcessor>
+XyzWriterPointsProcessor::FromDictionary(
+    common::LuaParameterDictionary* dictionary, PointsProcessor* next) {
```
Comment (damonkohler) [inline cartographer/io/xyz_writing_points_processor.cc]: const

File: cartographer/io/points_processor_pipeline_builder.cc [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  std::vector<std::unique_ptr<PointsProcessor>> pipeline;
+  pipeline.emplace_back(common::make_unique<NullPointsProcessor>());
```
Comment (damonkohler) [inline cartographer/io/points_processor_pipeline_builder.cc]: Why is this here?

File: cartographer/io/points_processor_pipeline_builder.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  PointsProcessorPipelineBuilder& operator=(
+      const PointsProcessorPipelineBuilder&) = delete;
+
+  static PointsProcessorPipelineBuilder* instance();
+
+  template <typename PointsProcessorType>
+  void Register() {
+    instance()->RegisterType(
+        PointsProcessorType::kConfigurationFileActionName,
+        [](common::LuaParameterDictionary* dictionary,
```
Comment (damonkohler) [inline cartographer/io/points_processor_pipeline_builder.h]: const

File: cartographer/io/points_processor_pipeline_builder.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+class PointsProcessorPipelineBuilder {
+ public:
+  PointsProcessorPipelineBuilder(const PointsProcessorPipelineBuilder&) =
+      delete;
+  PointsProcessorPipelineBuilder& operator=(
+      const PointsProcessorPipelineBuilder&) = delete;
+
+  static PointsProcessorPipelineBuilder* instance();
+
+  template <typename PointsProcessorType>
```
Comment (damonkohler) [inline cartographer/io/points_processor_pipeline_builder.h]: Why use a template instead of simply extending the PointsProcessor interface?

Comment (SirVer) [inline cartographer/io/points_processor_pipeline_builder.h]: GetName() would need to be a virtual static member function. that does not exist.
